src/main/java/htsjdk/samtools/util/ParallelBlockCompressedOutputStream.java

```diff
@@ -0,0 +1,310 @@
+        submitSpillTask();
+        return super.getFilePointer();
+    }
+
+    /**
+     * Submit new deflate task of the current uncompressed byte buffer.
+     */
+    @Override
+    protected int deflateBlock() {
+        if (numUncompressedBytes == 0) {
+            return 0;
+        }
+        submitDeflateTask();
+        numUncompressedBytes = 0;
+
+        if (compressedBlocksInFuture.size() == BLOCKS_PACK_SIZE) {
+            submitSpillTask();
+        }
+
+        return 0;
+    }
+
+    private void submitDeflateTask() {
+        UncompressedBlock uncompressedBlock = new UncompressedBlock(
+                Arrays.copyOf(this.uncompressedBuffer, numUncompressedBytes),
+                numUncompressedBytes
+        );
+        compressedBlocksInFuture.add(
+                CompletableFuture.supplyAsync(new BlockZipper(uncompressedBlock), gzipExecutorService)
+        );
+    }
+
+    private void submitSpillTask() {
+        writeBlocksTask.join();
+        writeBlocksTask = CompletableFuture.runAsync(new BlocksSpiller(compressedBlocksInFuture), gzipExecutorService);
+        compressedBlocksInFuture = new ArrayList<>(BLOCKS_PACK_SIZE);
+    }
+
+    /**
+     * Inner class which represent task for spilling block which was passed
+     * */
+    private class BlocksSpiller implements Runnable {
+
+        private List<Future<CompressedBlock>> compressedBlocks;
+
+        BlocksSpiller(List<Future<CompressedBlock>> compressedBlocks) {
+            this.compressedBlocks = compressedBlocks;
+        }
+
+        @Override
+        public void run() {
+            for (Future<CompressedBlock> compressedBlockFuture : compressedBlocks) {
+                try {
+                    CompressedBlock compressedBlock = compressedBlockFuture.get();
+                    mBlockAddress += writeGzipBlock(
+                            compressedBlock.compressedBuffer,
+                            compressedBlock.compressedSize,
+                            compressedBlock.bytesToCompress,
+                            compressedBlock.crc32Value
+                    );
+                } catch (InterruptedException | ExecutionException e) {
+                    throw new RuntimeIOException("Enable to write Gzip block", e);
+                }
+            }
+        }
+    }
+
+
+    /**
+     * Inner class which represent task for zipping block which was passed
+     * */
+    private class BlockZipper implements Supplier<CompressedBlock> {
+
+        private UncompressedBlock block;
+
+        BlockZipper(UncompressedBlock block) {
+            this.block = block;
+        }
+
+        @Override
+        public CompressedBlock get() {
+            return compressBlock(block);
+
+        }
+
+        private CompressedBlock compressBlock(UncompressedBlock uncompressedBlock) {
+            Deflater noCompressionDeflater = deflaterFactory.makeDeflater(Deflater.NO_COMPRESSION, true);
```

Comment:
All local variables in this method could be final. Also, there is no reason to initialise the `noCompressionDeflater` in here. We can initialise it only if we need to fall back to it.